lib/documents.js

```diff
@@ -3661,6 +3660,112 @@ function finishTransformAll(jobState) {
+ *  strings.
+ * @param {Integer} [batchSize] - The number of documents in each request where the inputKind is 'string'.
+ * @param {object} [concurrentRequests] - JavaScript object literal that controls the maximum number of concurrent
+ *  requests that can be pending at the same time. Valid keys are "multipleOf" and "multiplier".
+ * @param {string|mixed[]}  [transform] - the name of a transform extension to apply to each document or an array
+ *   with the name of the transform extension and an object of parameter values; the transform must have been
+ *   installed using the {@link transforms#write} function.
+ * @param {replace|ignore} [transformStrategy] - Controls whether the output from the transform replaces the content at
+ *  the document URI. The ignore strategy allows the transform to persist output in the database without changing
+ *  the original document.
+ * @param {function(progress, documents)} [onBatchSuccess] - A callback that can notify other systems about batches
+ *  transformed successfully.
+ */
+Documents.prototype.queryToTransformAll = function queryToTransformAllDocuments(query, options) {
+  return queryToTransformAllDocumentsImpl.call(
+      this, query, options
+  );
+};
+
+function queryToTransformAllDocumentsImpl(query, jobOptions){
+  let queryAllJobOptions = {};
+  let transformAllJobOptions = {};
+
+  if(!jobOptions){
+    jobOptions = {};
+  }
+  queryAllJobOptions.batchSize = (jobOptions.batchSize) ? jobOptions.batchSize : 250;
+
+  if(queryAllJobOptions.batchSize > 1){
+    transformAllJobOptions.inputKind = 'array';
+  } else {
+    transformAllJobOptions.batchSize = queryAllJobOptions.batchSize;
+  }
+
+  if(jobOptions.queryBatchMultiple) {
+    queryAllJobOptions.queryBatchMultiple = jobOptions.queryBatchMultiple;
+  }
+
+  if (jobOptions.consistentSnapshot) {
+    queryAllJobOptions.consistentSnapshot = jobOptions.consistentSnapshot;
+  }
+
+  if(jobOptions.onInitialTimestamp){
+    queryAllJobOptions.onInitialTimestamp = jobOptions.onInitialTimestamp;
+  }
+
+  if (jobOptions.onCompletion) {
+    transformAllJobOptions.onCompletion = jobOptions.onCompletion;
+  }
+  if (jobOptions.onBatchError) {
+    transformAllJobOptions.onBatchError = jobOptions.onBatchError;
+  }
+  if (jobOptions.concurrentRequests) {
+    transformAllJobOptions.concurrentRequests = jobOptions.concurrentRequests;
+  }
+  if(jobOptions.transform){
+    transformAllJobOptions.transform = jobOptions.transform;
+  }
+  if(jobOptions.transformStrategy){
+    transformAllJobOptions.transformStrategy = jobOptions.transformStrategy;
+  }
+  if(jobOptions.onBatchSuccess){
+    transformAllJobOptions.onBatchSuccess = jobOptions.onBatchSuccess;
+  }
+
+  queryAllJobOptions.queryToOptions = transformAllJobOptions;
+
+  const docInstance = this;
+  let queryAllStream = queryAllDocumentsImpl.call(docInstance, query, queryAllJobOptions);
+  let pipelined = false;
+
+  queryAllStream.on('data', function(item){
```

Comment:
I'm still wondering whether there is an idiomatic approach in Node.js that can pause the transformAllStream until the queryAllStream is ready to send data so the streams can be piped initially instead of waiting for the first data event to pipe.

But, I think there's already an issue to investigate that.  This would just be another potential case for refactoring to add  to that issue.